docs/389ds/design/worker-threads.md

```diff
@@ -0,0 +1,345 @@
+  - Release the global mutex
+  - Walk closing list:
+    - Get thread mutex
+    - Set thread closing flags
+    - Wake up the thread
+    - Release thread mutex
+  - Walk closing list:
+    - Join the thread
+    - Unlink thread from closing list
+    - Free thread context
+
+### Monitoring
+
+The following data should be displayed:
+
+| Data                        | Attribute name | Description                                                   |
+| --------------------------- | -------------- | ------------------------------------------------------------- |
+| waiting_thread size         | waitingthreads | Number of worker threads that are waiting for jobs            |
+| busy_thread size            | busythreads    | Number of worker threads that are processing on  jobs         |
+| busy_thread high water mark | maxbusythreads | Highest number of worker threads that are processing on  jobs |
+| waiting_job size            | waitingjobs    | Size of the waiting job queue                                 |
+| waiting_job high water mark | maxwaitingjobs | Highest size of the waiting job queue                         |
+
+### Test
+
+An automatic test can be done with the following steps:
+
+- create an instance
+- get the monitoring results
+- check that waitingthreads+busythreads == configured number of threads
+- run pstack of server and check that number of connection_threadmain is the expected one 
+- increase the number of threads
+- check that waitingthreads+busythreads == configured number of threads
+- run pstack of server and check that number of connection_threadmain is the expected one
+- decrease the number of threads
+- check that waitingthreads+busythreads == configured number of threads
+- run pstack of server and check that number of connection_threadmain is the expected one
+
+### Considered Alternatives
+
+- improving old worker thread algorithm to be able to change the number of threads dynamically. (Rejected because a prototype showed that on my laptop, when using trivial jobs that atommically increment a counter, the proposed algorithm was 6 time faster than the older)
+- Adjusting automatically the number of threads to the load. (Rejected because an hard limit is still needed to limit the resource use if an operation blocks the server for some time - and with the new algorithm useless configured threads should not have a noticeable impact (except on the virtual memory footprint)
+- while pushing job in waiting_job queue:
+  - Do not limit list size (Rejected because of the risk of exhausting 
+     the system resources)
+  - alloc job list element (while keeping the global lock or use pre-alloced elements as in current code). Rationnal: This queuing occurs when incomming job load is higher that what the working threads can handle, so the priority is to let the worker threads picks new job as fast as possible and we can slow the listening threads.
+- Using NSPR thread / mutex / condvar or using pthreads
```

Comment:
So what is the decision ? use pthread or nspr mutex/condvar ?

Reply:
FYI: final decision is not done yet:
 Right now, I plan to use pthreads (to limit the overhead) but I may change my mind if there are too many impact on the current code (I am a bit wary about the thread specific data handling (not sure that nspr function does work on pthreads) for the mutex/condvar it is more simple (pthreads are more efficients)

Reply:
Done (cf line 227):
 Ok, now that there is a working prototype, I have the answer: The server seems happy with pthreads, so lets use it because it has less overhead than nspr